src/ophyd_async/core/device.py

```diff
@@ -83,12 +91,21 @@ async def connect(self, sim: bool = False, timeout: float = DEFAULT_TIMEOUT):
         timeout:
             Time to wait before failing with a TimeoutError.
         """
-        coros = {
-            name: child_device.connect(sim, timeout=timeout)
-            for name, child_device in self.children()
-        }
-        if coros:
-            await wait_for_connection(**coros)
+        previous_connect_ok:bool = (
+            self._connect_task is not None
+            and self._connect_task.done()
+            and not self._connect_task.exception()
+        )
+        if force_reconnect or not previous_connect_ok:
+            # Kick off a connection
+            coros = {
+                name: child_device.connect(sim, timeout=timeout)
+                for name, child_device in self.children()
+            }
+            self._connect_task = asyncio.create_task(wait_for_connection(**coros))
+        # Wait for it to complete
+        assert self._connect_task, "Connect task not created, this shouldn't happen"
+        await self._connect_task
```

Comment:
```suggestion
        if force_reconnect or not self._previous_connect_ok:
            # Kick off a connection
            coros = {
                name: child_device.connect(sim, timeout=timeout)
                for name, child_device in self.children()
            }
            connect_task = asyncio.create_task(wait_for_connection(**coros))
            # Wait for it to complete
            await connect_task
            self._previous_connect_ok = not connect_task.exception()
```

Since we're awaiting the task within the method anyway, I think it's probably nicer to do this...